### ISSUE TITLE
Group speakers

### ISSUE THREAD

**Issue by gonzalorui** [Strixx76/samsungwam#4]
Hello, thanks for your work.  Wait a long time for this integration.  Are you planning to implement a function to group speakers?

**Comment by Strixx76** [comment]
Yes. I am actually working on it right now. I have some final testing to do before I release it. Hopefully it will be available the within a week or two.

**Comment by Strixx76** [comment]
This feature is now implemented but I need some more weeks to test it. It will only work with normal multiroom grouping, and not with surround grouping for TV-sound. Hopefully a new release will be available before the end of the month.

**Comment by gonzalorui** [comment]
Hi, can the update be downloaded? I don't see it....

**Comment by Strixx76** [comment]
As I wrote I need some more time for final testing, but hopefully a new release will be available before the end of the month. Then you will be able to download it.

**Comment by DaTiezl** [comment]
Thanks for all your effort in the limited time you have available. I just downloaded your RC with this feature. Can you give us some configuration options. I would like to help you with testing!

**Comment by Strixx76** [comment]
I am about to update README.md. But if you want to try out the grouping I recommend installing [Mini Media Player card](https://github.com/kalkih/mini-media-player#speaker-group-management).
Hopefully the RC works out of the box, but I made some last minute changes before releasing it, that I haven't actually been able to test my self yet. Those changes are to be able to fix dependency problem later on. And I also needed to do a release to be able to get the integration in to HACS default library of integrations. If everything works well, there will be a ordinary release later this week.

**Comment by DaTiezl** [comment]
I have been using the integration for some time now and also the code above to be able to group speakers. Unfortunately I cannot say that the control of my speakers works very well. I still often experience problems with the integration. Speakers that do not turn on or stay on but also the grouping does not work well. Regularly the other speakers remain greyed-out and I have to use the Samsung app itself to group.

Maby it has something to do with my configuration..

My daily errors in log:
- All speakers in group is not available in Home Assistant
- Error while getting speaker data (ConnectionError: Could not connect to speaker)
- Could not connect to speaker: ([speakername]@[ip]) (pywam.lib.exceptions.ApiCallTimeoutError: No response from speaker)

**Comment by Strixx76** [comment]
@DaTiezl please create a new issue and provide some more info. This is for implementing the grouping feature which is now implemented and will be closed as soon as I release a new version.

**Comment by Strixx76** [comment]
I have found some issues with how I handle the grouping calls to the speakers when trying out a new card [(maxi-media-player)](https://github.com/punxaphil/maxi-media-player) to control my speakers. And  while I did some more testing I realized that this also effect the mini-media-player card. This will show if you try to "edit" a group or if you add or remove more than one speaker at a time. I am in the process of rewriting the logic, and hope to be able to release a more stable version before the end of the week.

**Comment by DaTiezl** [comment]
Sorry Daniel for neglecting to create the issue and not providing more info. As soon as you release the new version I will test more extensively with you and provide info.

**Comment by Strixx76** [comment]
There is now a new release candidate to test. v0.1.0rc7 that hopefully have solved all issues with grouping. Here is a link explaining how to download latest rc: [Downloading a specific version of a repository](https://www.hacs.xyz/docs/use/repositories/dashboard/#downloading-a-specific-version-of-a-repository)

**Comment by DaTiezl** [comment]
So far everything is fine. I will continue testing. 

What I notice is that I can no longer select a source of a when the speaker is in slave mode. Perhaps it is an option to keep the other sources selectable and then degroup the speaker in case of a selection other than the master speaker?

**Comment by ingulacrist** [comment]
@DaTiezl It's the same in the Samsung app: you cannot select a source for a grouped (slave) speaker or LinkMate. Therefore this would be more advanced functionality than the official app offers.

**Comment by DaTiezl** [comment]
That's true. It would be an advanced feature. Ungrouping a single speaker from the group is also possible using this integration and not in the Samsung app so there are some additional features already.

**Comment by ingulacrist** [comment]
Technically, ungrouping a single speaker is possible in the app by editing the group and unchecking the one speaker you want to ungroup (or just push one of the physical input buttons on the speaker itself and it will drop out of the group).

**Comment by Strixx76** [comment]
@DaTiezl In the latest rc I changed so that the selected source for slaves shows the name of the master speaker when they are grouped. Before it showed "WiFi". But you can't change it, and you could not before either.

It is not hard to have the sources show and ungroup the speaker if you choose another source. But I am not so sure about if that would be very user friendly. That would mean that for them master some sources will cause the group to be deleted and others not. Which could be quite confuseing.

For the slaves all other than select would mean that the single speaker would leave the group. This could be more predictable if I also add all Samsung speakers (master and ungrouped, but not slaves) to the source list so that you can choose another speaker as source. The problem with this approach would be a very long list of sources if you have a couple of speakers. The upside would be that you would be able to group speakers with default media player cards.

I will have to think about it and perhaps implement it on next release after this. Right now I would like to focus on releasing a non rc version and get the integration added to HACS default.

**Comment by DaTiezl** [comment]
I understand your argument and I think you might be right. Just leave it as it is then.

**Comment by Strixx76** [comment]
[v0.1.0](https://github.com/Strixx76/samsungwam/releases/tag/v0.1.0)
